embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,881 @@
+    /// mode is intended for application threads on a single core, not interrupts.
+    /// As such, only one task at a time can acquire a resource and so this
+    /// channel avoids all locks. To create one:
+    ///
+    /// ``` no_run
+    /// use core::cell::UnsafeCell;
+    /// use embassy::util::mpsc;
+    /// use embassy::util::mpsc::{Channel, WithThreadModeOnly};
+    ///
+    /// // Declare a bounded channel of 3 u32s.
+    /// let mut channel = UnsafeCell::new(Channel::<WithThreadModeOnly, u32, 3>::with_thread_mode_only());
+    /// // once we have a channel, obtain its sender and receiver
+    /// let (sender, receiver) = mpsc::split(&channel);
+    /// ```
+    pub const fn with_thread_mode_only() -> Self {
+        let mutex = ThreadModeMutex::new(());
+        let state = ChannelState::new();
+        Channel { mutex, state }
+    }
+}
+
+impl<M, T, const N: usize> ChannelLike<T> for Channel<M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    fn try_recv(&mut self) -> Result<T, TryRecvError> {
+        let state = &mut self.state;
+        self.mutex.lock(|_| {
+            if !state.closed {
+                if state.read_pos != state.write_pos || state.full {
+                    if state.full {
+                        state.full = false;
+                        state.senders_waker.take().map(|w| w.wake());
+                    }
+                    let message =
+                        unsafe { (state.buf[state.read_pos]).assume_init_mut().get().read() };
+                    state.read_pos = (state.read_pos + 1) % state.buf.len();
+                    Ok(message)
+                } else if !state.closing {
+                    Err(TryRecvError::Empty)
+                } else {
+                    state.closed = true;
+                    state.senders_waker.take().map(|w| w.wake());
+                    Err(TryRecvError::Closed)
+                }
+            } else {
+                Err(TryRecvError::Closed)
+            }
+        })
+    }
+
+    fn try_send(&mut self, message: T) -> Result<(), TrySendError<T>> {
+        let state = &mut self.state;
+        self.mutex.lock(|_| {
+            if !state.closed {
+                if !state.full {
+                    state.buf[state.write_pos] = MaybeUninit::new(message.into());
+                    state.write_pos = (state.write_pos + 1) % state.buf.len();
+                    if state.write_pos == state.read_pos {
+                        state.full = true;
+                    }
+                    state.receiver_waker.take().map(|w| w.wake());
+                    Ok(())
+                } else {
+                    Err(TrySendError::Full(message))
+                }
+            } else {
+                Err(TrySendError::Closed(message))
+            }
+        })
+    }
+
+    fn close(&mut self) {
+        let state = &mut self.state;
+        self.mutex.lock(|_| {
+            state.receiver_waker.take().map(|w| w.wake());
```

Comment:
I think it's a bit weird to use map just for the side-effects, maybe a `if let` is better here ?

Reply:
Changed. Thanks.